src/main/java/io/confluent/copycat/hdfs/TopicPartitionWriter.java

```diff
@@ -0,0 +1,461 @@
+/**
+ * Copyright 2015 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ **/
+
+package io.confluent.copycat.hdfs;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.copycat.errors.CopycatException;
+import org.apache.kafka.copycat.errors.IllegalWorkerStateException;
+import org.apache.kafka.copycat.sink.SinkRecord;
+import org.apache.kafka.copycat.sink.SinkTaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import io.confluent.copycat.avro.AvroData;
+import io.confluent.copycat.hdfs.partitioner.Partitioner;
+
+public class TopicPartitionWriter {
+  private static final Logger log = LoggerFactory.getLogger(TopicPartitionWriter.class);
+  private WAL wal;
+  private Map<String, String> tempfiles;
+  private Map<String, RecordWriter> writers;
+  private TopicPartition tp;
+  private Partitioner partitioner;
+  private String url;
+  private String topicsDir;
+  private State state;
+  private Queue<SinkRecord> buffer;
+  private boolean recovered;
+  private Storage storage;
+  private SinkTaskContext context;
+  private int recordCounter;
+  private int flushSize;
+  private RecordWriterProvider writerProvider;
+  private Configuration conf;
+  private AvroData avroData;
+  private Set<String> appended;
+  private long offset = -1;
+  private Map<String, Long> startOffsets;
+  private Map<String, Long> offsets;
+  private long timeoutMs;
+  private long failureTime;
+
+  private Pattern pattern = Pattern.compile("\\w+_\\d+_\\d+_\\d+");
```

Comment:
I saw this pattern before, any place we could store it just once? Especially since I think it needs tweaking, so they are at risk of getting out of sync with each other.